pennylane_lightning/src/util/Util.hpp

```diff
@@ -915,6 +915,80 @@ auto linspace(T start, T end, size_t num_points) -> std::vector<T> {
     return data;
 }
 
+/**
+ * @brief Determines the indices that would sort an array.
+ *
+ * @tparam T vector data type.
+ * @param arr Array to be inspected.
+ * @return a vector with indices that would sort the array.
+ */
+template <typename T>
+inline auto sorting_indexes(const T &arr, size_t length)
+                            -> std::vector<size_t>  {
+  std::vector<size_t> indexes(length);
+  iota(indexes.begin(), indexes.end(), 0);
+
+  // indexes will be sorted in accordance to the array provided.
+  sort(indexes.begin(), indexes.end(),
+       [&arr](size_t i1, size_t i2) {return arr[i1] < arr[i2];});
+
+  return indexes;
+}
+
+/**
+ * @brief Determines the indices that would sort a vector.
+ *
+ * @tparam T array data type.
+ * @param vec Vector to be inspected.
+ * @return a vector with indices that would sort the vector.
+ */
+template <typename T>
+inline auto sorting_indexes(const std::vector<T> &vec) 
+                            -> std::vector<size_t> {
+  return sorting_indexes(vec.data(), vec.size());
+}
+
+/**
+ * @brief Determines a new index for a transposed tensor stored linearly.
+ *  This function assumes each axis will have a length of 2 (|0>, |1>).
+ *
+ * @param ind index before.
+ * @param new_axes new axes distribution.
+ * @return unsigned int with the new transposed index.
+ */
+inline auto transposed_state_index (  size_t ind, 
+                                        const std::vector<size_t> &new_axes)
+                                    -> size_t {
+    size_t new_index=0;
+    for (std::vector<size_t>::const_reverse_iterator axis = new_axes.rbegin(); 
+            axis != new_axes.rend(); ++axis){
+        new_index+=(ind%2)*pow(2,(new_axes.size()-1)-(*axis));
```

Comment:
```suggestion
        new_index += ((ind%2) << ((new_axes.size()-1)-(*axis)));
```

Powers of 2 can be faster (usually) using l-shift operations. I think this should be the same (but please check it for me).

Reply:
I will. Nice suggestion. Thanks.